extensions/pkg/controller/dnsrecord/reconciler.go

```diff
@@ -0,0 +1,217 @@
+const RequeueAfter = 2 * time.Second
+
+type reconciler struct {
+	logger   logr.Logger
+	actuator Actuator
+
+	client        client.Client
+	reader        client.Reader
+	statusUpdater extensionscontroller.StatusUpdater
+}
+
+// NewReconciler creates a new reconcile.Reconciler that reconciles
+// dnsrecord resources of Gardener's `extensions.gardener.cloud` API group.
+func NewReconciler(actuator Actuator) reconcile.Reconciler {
+	logger := log.Log.WithName(ControllerName)
+
+	return extensionscontroller.OperationAnnotationWrapper(
+		func() client.Object { return &extensionsv1alpha1.DNSRecord{} },
+		&reconciler{
+			logger:        logger,
+			actuator:      actuator,
+			statusUpdater: extensionscontroller.NewStatusUpdater(logger),
+		},
+	)
+}
+
+func (r *reconciler) InjectFunc(f inject.Func) error {
+	return f(r.actuator)
+}
+
+func (r *reconciler) InjectClient(client client.Client) error {
+	r.client = client
+	r.statusUpdater.InjectClient(client)
+	return nil
+}
+
+func (r *reconciler) InjectAPIReader(reader client.Reader) error {
+	r.reader = reader
+	return nil
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	dns := &extensionsv1alpha1.DNSRecord{}
+	if err := r.client.Get(ctx, request.NamespacedName, dns); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	cluster, err := extensionscontroller.GetCluster(ctx, r.client, dns.Namespace)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if extensionscontroller.IsFailed(cluster) {
+		r.logger.Info("Stop reconciling DNSRecord of failed Shoot", "namespace", request.Namespace, "name", dns.Name)
+		return reconcile.Result{}, nil
+	}
+
+	operationType := gardencorev1beta1helper.ComputeOperationType(dns.ObjectMeta, dns.Status.LastOperation)
+
+	switch {
+	case extensionscontroller.IsMigrated(dns):
+		return reconcile.Result{}, nil
+	case operationType == gardencorev1beta1.LastOperationTypeMigrate:
+		return r.migrate(ctx, dns, cluster)
+	case dns.DeletionTimestamp != nil:
+		return r.delete(ctx, dns, cluster)
+	case dns.Annotations[v1beta1constants.GardenerOperation] == v1beta1constants.GardenerOperationRestore:
+		return r.restore(ctx, dns, cluster)
+	default:
+		return r.reconcile(ctx, dns, cluster, operationType)
+	}
+}
+
+func (r *reconciler) reconcile(ctx context.Context, dns *extensionsv1alpha1.DNSRecord, cluster *extensionscontroller.Cluster, operationType gardencorev1beta1.LastOperationType) (reconcile.Result, error) {
+	if err := controllerutils.EnsureFinalizer(ctx, r.reader, r.client, dns, FinalizerName); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err := r.statusUpdater.Processing(ctx, dns, operationType, "Reconciling the dnsrecord"); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	r.logger.Info("Starting the reconciliation of dnsrecord", "dnsrecord", dns.Name)
```

Comment:
Add a namespace here?
```suggestion
	r.logger.Info("Starting the reconciliation of dnsrecord", "dnsrecord", client.ObjectKeyFromObject(dns).String())
```

Reply:
Used `kutil.ObjectName` everywhere, used lowercase `dnsrecord` (as opposed to `DNSRecord`) in all messages, and also improved the phrasing of a few messages. If you like it, I could make these changes across all extension reconcilers in another PR (they are suffering from the same minor bugs / inconsistencies).